### DISCUSSION TITLE
Can transcript_words.txt file contain only unique and sorted lines ?

### DISCUSSION THREAD

**Discussion by ChrystianKacki** [k2-fsa/icefall#1615] [Q&A] [answered]
In one of the final stages of scripts `egs/.../ASR/prepare.sh`, there is creating of `transcript_words.txt` file.

After opening this `transcript_words.txt` file, I see that it contains duplicated lines.

Can `transcript_words.txt` contain only unique lines ?
Can lines in `transcript_words.txt` be sorted ?

**Comment by JinZr** [comment]
hi,

may i ask which recipe are you using? 

best
jin

**Reply by ChrystianKacki** [reply]
Hi,
I am using my own recipe based on Common Voice recipe, with combined MLS and VoxPopuli.
I used `egs/commonvoice/ASR/prepare.sh` to create my own version.

**Comment by JinZr** [comment]
i see, are you using the ``jq`` command to grep text lines from the jsonl.gz format cutset?

best
jin

**Reply by ChrystianKacki** [reply]
Yes, it is in line `#L342` of `egs/commonvoice/ASR/prepare.sh`

**Reply by ChrystianKacki** [reply]
Also in `prepare.sh` I set `use_validated=true`. I checked transcripts file `validated.tsv` of newest CV 17.0 dataset for Polish language and there are repeating sentences for different speakers.

**Comment by JinZr** [comment]
i see, the `transcript_words.txt` is only used for BPE and ngram LM training, so I dont think several duplicated sentences or sorting it would cause a huge problem. 

I’m a little bit concerned about the repeating sentences of different speakers, would you do me a favor and check if those sentences really exist in the training set, and there are corresponding audio files matching with those text?

best
jin

**Comment by JinZr** [comment]
also are there a lot of duplications presented in the `validated.tsv` file?

best
jin

**Comment by ChrystianKacki** [comment]
I prepared a file with repeating sentences count in Polish CV validated training set.
I checked it and all the sentences have corresponding and unique audio files.
Please see it: [common_voice-pl_cuts_validated_raw-line_count.txt](https://github.com/k2-fsa/icefall/files/15208450/common_voice-pl_cuts_validated_raw-line_count.txt)

**Comment by JinZr** [comment, marked as answer]
i see, there really is a lot of duplication in the transcript. 

i suppose you can try manually removing all duplicated sentences, this would at least benefit the performance of your ngram model.

best
jin

**Reply by ChrystianKacki** [reply]
Great. Can those unique sentences be sorted ?
I'm asking because in Bash it's easier to sort a file before removing duplicates.
Best, Chrystian

**Comment by JinZr** [comment]
yes, sorting doesn’t interfere with BPE and ngram LM training

best
jin

**Reply by ChrystianKacki** [reply]
Thanks!
Best, Chrystian
